tenseal/tensors/abstract_tensor.py

```diff
@@ -0,0 +1,152 @@
+"""
+"""
+
+import tenseal as ts
+from abc import ABC
+
+
+class AbstractTensor(ABC):
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        native_type = getattr(ts._ts_cpp, self.__class__.__name__)
+        if not isinstance(value, native_type):
+            raise TypeError(f"value must be of type {native_type}")
+        self._data = value
+
+    def copy(self):
+        return self._wrap(data=self.data.copy())
+
+    def __copy__(self):
+        return self.copy()
+
+    def context(self):
+        return ts.Context._wrap(self.data.context())
+
+    @property
+    def shape(self):
+        return self.data.shape()
+
+    @classmethod
+    def load(cls, context, data):
+        """
+        Constructor method for the tensor object from a serialized protobuffer.
+        Args:
+            context: a context, holding the encryption parameters and keys.
+            data: the serialized protobuffer.
+        Returns:
+            Tensor object.
+        """
+        if isinstance(context, ts.Context) and isinstance(data, bytes):
+            native_type = getattr(ts._ts_cpp, cls.__name__)
+            return cls._wrap(native_type(context.data, data))
+
+        raise TypeError(
+            "Invalid input types context: {} and vector: {}".format(type(context), type(data))
+        )
+
+    def serialize(self):
+        return self.data.serialize()
+
+    @classmethod
+    def _wrap(cls, data):
+        return cls(data=data)
+
+    def _decrypt(self, secret_key=None):
+        if secret_key is None:
+            return self.data.decrypt()
+        elif isinstance(secret_key, ts.SecretKey):
+            return self.data.decrypt(secret_key.data)
+        else:
+            raise TypeError(f"incorrect type: {type(secret_key)} != SecretKey")
+
+    @classmethod
+    def _get_operand(cls, other, dtype="float"):
+        if isinstance(other, (int, float)):
+            return other
+        elif isinstance(other, (cls, ts.PlainTensor)):
+            return other.data
+        else:
+            try:
+                other = ts.plain_tensor(other, dtype=dtype)
+                other = other.data
+            except TypeError:
+                raise TypeError(f"can't operate with object of type {type(other)}")
+            return other
+
+    def __add__(self, other):
```

Comment:
Do we need to keep the python arithmetics overloads in pybind in this case?

Reply:
Currently yes, because I'm usig the arithmetic operators (+,-,*) in the implementation, but yeah I can remove them and start using functions